test/pylib/cpp/item.py

```diff
@@ -0,0 +1,141 @@
+        filename, line_num = failure.get_file_reference()
+        return ReprFileLocation(filename, line_num, "C++ failure")
+
+    def toterminal(self, tw: TerminalWriter) -> None:
+        for index, failure in enumerate(self.failures):
+            for line, markup in failure.get_lines():
+                markup_params = {m: True for m in markup}
+                tw.line(line, **markup_params)
+
+            location = self._get_repr_file_location(failure)
+            location.toterminal(tw)
+
+            if index != len(self.failures) - 1:
+                tw.line(self.failure_sep, cyan=True)
+
+
+class CppItem(pytest.Item):
+    """
+    Represents a single test function in the file.
+    """
+    facade = None
+
+    def __init__(self, *, executable: Path, facade: TestFacade, mode: str, test_name: str, arguments: Sequence[str],
+                 file_name, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.facade = facade
+        self.executable = executable
+        self.mode = mode
+        self.file_name = file_name
+        self.test_name = test_name
+        self._arguments = arguments
+        self.add_marker(pytest.mark.cpp)
+
+    @property
+    def nodeid(self) -> str:
+        return self._nodeid
+
+    @nodeid.setter
+    def nodeid(self, nodeid: str) -> None:
+        self._nodeid = nodeid
+
+    def runtest(self) -> None:
+
+        failures, output = self.facade.run_test(str(self.executable), self.test_name, self.mode, self.file_name,
+                                                self._arguments)
+        # Report the c++ output in its own sections
+        self.add_report_section("call", "c++", output)
+
+        if failures:
+            raise CppFailureError(failures)
+
+    def repr_failure(  # type:ignore[override]
+            self, excinfo: pytest.ExceptionInfo[BaseException], **kwargs: Any) -> str | TerminalRepr | CppFailureRepr:
+        if isinstance(excinfo.value, CppFailureError):
+            return CppFailureRepr(excinfo.value.failures)
+        return pytest.Item.repr_failure(self, excinfo)
+
+    def reportinfo(self) -> tuple[Any, int, str]:
+        return self.path, 0, self.name
+
+
+class CppFile(pytest.File):
+    """
+    Represents the C++ test file with all necessary information for test execution
+    """
+    def __init__(self, *, no_parallel_run: bool = False, modes: list[str], disabled_tests: dict[str, set[str]],
+                 run_id=None, facade: TestFacade, arguments: Sequence[str], parameters: list[str] = None,
+                 **kwargs: Any, ) -> None:
+        super().__init__(**kwargs)
+        self.facade = facade
+        self.modes = modes
+        self.run_id = run_id
+        self.disabled_tests = disabled_tests
+        self.no_parallel_run = no_parallel_run
+        self.parameters = parameters
+        self._arguments = arguments
+
+    def collect(self) -> Iterator[CppItem]:
+        project_root = find_repo(self.path)
```

Comment:
in pytest the collection starts from the current dir, why do you need to go to the repo here?

Reply:
It's needed in line 118 to find the build directory and create a full path to the compiled binary.